src/storage/mvcc/txn.rs

```diff
@@ -205,9 +219,103 @@ impl<S: Snapshot> MvccTxn<S> {
         Ok(())
     }
 
+    pub fn acquire_pessimistic_lock(
+        &mut self,
+        key: Key,
+        primary: &[u8],
+        for_update_ts: u64,
+        options: &Options,
+    ) -> Result<()> {
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.into_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                return Err(Error::LockTypeNotMatch {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                    pessimistic: false,
+                });
+            }
+            MVCC_DUPLICATE_CMD_COUNTER_VEC.pessimistic_lock.inc();
+            return Ok(());
+        }
+
+        if let Some((commit_ts, _)) = self.reader.seek_write(&key, u64::max_value())? {
+            // The isolation level of pessimistic transactions is RC. `for_update_ts` is
+            // the commit_ts of the data this transaction read. If exists a commit version
+            // whose commit timestamp is larger than current `for_update_ts`, the
+            // transaction should retry to get the latest data.
+            if commit_ts > for_update_ts {
+                MVCC_CONFLICT_COUNTER.pessimistic_lock_conflict.inc();
+                return Err(Error::WriteConflict {
+                    start_ts: self.start_ts,
+                    // TiDB will try to extract the latest `commit_ts` from WriteConflict
+                    // error to avoid getting a timestamp from tso. So we set
+                    // `conflict_start_ts` to `commit_ts` here.
+                    conflict_start_ts: commit_ts,
+                    conflict_commit_ts: commit_ts,
+                    key: key.into_raw()?,
+                    primary: primary.to_vec(),
+                });
+            }
+
+            // Handle rollback.
+            // If the commit timestamp is equal to transaction's start timestamp, the
+            // lock is already rollbacked.
+            if commit_ts == self.start_ts {
+                return Err(Error::Rollbacked {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                });
+            }
+            // If `commit_ts` we seek is already before `start_ts`, the rollback must not exist.
+            if commit_ts > self.start_ts {
+                if let Some((commit_ts, _)) = self.reader.seek_write(&key, self.start_ts)? {
+                    // Rollback's commit_ts is equal to start_ts
```

Comment:
Please also check the commit type.

Reply:
Here is the RFC for the optimization on commit_ts.
https://github.com/tikv/rfcs/pull/25/files?short_path=6db5b0b#diff-6db5b0b7172e00123f1a97f6026ab290
To adopt this optimization in the future, please note that, when the latest commit_ts = the current start_ts, it doesn't imply that the commit_ts belongs to a `rollback` of the same transaction.
It's also ok if you have some ideas about improving the design of the commit_ts optimization. At least I think checking the commit type here is never a bad idea.

Reply:
If `write.start_ts == txn's start_ts && write.start_ts == commit_ts`, it must be a rollback record.Because there is no txns with the same `start_ts`.

Reply:
I guess you misunderstood what `seek_write` does. It finds `key`'s latest write record whose commit_ts <= start_ts.

Reply:
I knew it. One txn's `commit_ts` maybe equal to other txn's `start_ts`. But their `start_ts` must not be same. 
`Write` records txn's `start_ts`. if `write.start_ts == self.start_ts && write.start_ts == commit_ts`,  is it possible that the `Write` isn't a `rollbacked` record of the same txn?
I don't check `write.start_ts == self.start_ts` now. I think what i need to do is check it.

Reply:
Ok. And, I suggest adding an assertion for the commit type.

Reply:
yes